docs: add a tutorial blog for app runner vpc connectors (#3630)
aws/copilot-cli: site/content/blogs/apprunner-vpc.en.md

# AWS App Runner VPC connection with AWS Copilot

**Rafael Mosca**, Associate Solutions Architect, WWPS EMEA Central SA

The [AWS Copilot CLI](https://aws.github.io/copilot-cli/) is a tool that was [launched in 2020](https://aws.amazon.com/blogs/containers/introducing-aws-copilot/), and since then developers have been using it to build, manage, and operate Linux and Windows containers on [Amazon Elastic Container Service (Amazon ECS)](https://aws.amazon.com/ecs/), [AWS Fargate](https://aws.amazon.com/fargate/), and [AWS App Runner.](https://aws.amazon.com/apprunner/)

> AWS recently announced [VPC Support for AWS App Runner](https://aws.amazon.com/blogs/aws/new-for-app-runner-vpc-support/). This means that web applications and APIs that you deploy using [AWS App Runner](https://aws.amazon.com/apprunner/) can now communicate with databases running in services like [Amazon Relational Database Service (RDS)](https://aws.amazon.com/rds/), and other applications running on [Amazon Elastic Container Service (Amazon ECS)](https://aws.amazon.com/ecs/), [Amazon Elastic Kubernetes Service (EKS)](https://aws.amazon.com/eks/), [Amazon Elastic Compute Cloud (Amazon EC2)](https://aws.amazon.com/ec2/) that are hosted in an [Amazon Virtual Private Cloud (Amazon VPC)](https://aws.amazon.com/vpc/).

Reviewer: - Should we include a date for when App Runner announced VPC egress support?
- I wonder if we should also highlight that it's egress traffic (so requests within a VPC cannot reach the apprunner service yet)

Author: - Yeah maybe instead of saying _AWS recently announced VPC Support for AWS App Runner_ we can say: _In February 2022, AWS announced VPC Support for AWS App Runner_
- Regarding egress traffic only it should be written below :  "In summary, by creating a VPC connector, the App Runner service creates a uni-directional Elastic Network Interface (ENI) on the Customer-owned VPC that can be used for outbound traffic only (i.e. you cannot send any inbound requests to the ENI’s IP address)"

### Prerequisites

To build the architecture above, you need the following prerequisites:

- An [AWS account](https://aws.amazon.com/premiumsupport/knowledge-center/create-and-activate-aws-account/) and chosen a region in which you will deploy the infrastructure described above.
- Have the [AWS Copilot CLI](https://aws.github.io/copilot-cli/docs/getting-started/install/) installed on your local workstation.
- Properly configured [AWS credentials](https://aws.github.io/copilot-cli/docs/credentials/) using the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config) or with [environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html).
- [Docker](https://www.docker.com/products/docker-desktop) is installed and running on your local workstation.
- Cloned the sample repository
  - `git clone https://github.com/aws-rafams/apprunner-vpc-copilot`

### Create an application and environment

You will first need to create a logical group of related services, environments, and pipelines. In the AWS Copilot terminology this is called an _application_.

`copilot app init apprunner-vpc`

> Once you execute that command, Copilot will use a folder named `./copilot` to hold special YAML configuration files, called _manifests,_ that will enable you to easily deploy containerised applications on the AWS using Copilot.

Reviewer: nit: maybe like this?
```suggestion
Once you execute that command, Copilot will use a folder named `./copilot` to hold infrastructure-as-code YAML configuration files, called _manifests,_ that will enable you to easily deploy containerised applications on  AWS using Copilot.
```

> Your next step is to create an environment for the application where you will deploy your services. With AWS Copilot it is possible to create different environments that logically isolate the deployments of our applications in a very easy way. A common use case is to have a test environment and a separate production environment where applications are deployed only when they have been validated on the test environment. For the scope of this post, you will only deploy the services to a testing environment named _test_ that you create with the following command:

Reviewer: ```suggestion
Your next step is to create an environment for the application where you will deploy your services. With AWS Copilot it is possible to create different environments that logically isolate the deployments of our applications in separate accounts and regions. A common use case is to have a test environment and a separate production environment where applications are deployed only when they have been validated on the test environment. For the scope of this post, you will only deploy the services to a testing environment named _test_ that you create with the following command:
```

```
copilot env init \
    --app apprunner-vpc \
    --name test \
    --region 'eu-west-1' \
    --default-config
```

> Once you press enter on the command, you will be asked to select AWS credentials that will be used to create the necessary infrastructure to host our services. Once the credentials have been selected, Copilot will start to create the resources on your behalf. This process may take a while so stretch a bit while this process is completed.\n

Reviewer: huh does the \n work?
```suggestion
Once you press enter on the command, you will be asked to select AWS credentials that will be used to create the necessary infrastructure to host our services. Once the credentials have been selected, Copilot will start to create the resources on your behalf. This process may take a while so stretch a bit while this process is completed.  
```

For every environment you create, AWS Copilot will create a separate networking stack (VPC) and an ECS Cluster that won’t be used for this particular demo.
![AWS Copilot Groupings](./../assets/images/apprunner-vpc-blog/copilot-groupings.png)

### Create a service running on AWS App Runner

AWS Copilot gives us several abstractions that you can use to deploy different types of services. In this walkthrough example I use the Copilot pattern called [Request-Driven Web Service,](https://aws.github.io/copilot-cli/docs/manifest/rd-web-service/) which deploys an AWS App Runner service that autoscales based on incoming traffic and scales down to a baseline when there's no traffic. This option is more cost effective for HTTP services with sudden bursts in request volumes or low request volumes.

```
copilot svc init \
    --app apprunner-vpc  \
    --svc-type "Request-Driven Web Service" \
    --name demo-service\
    --port 5000 \
    --dockerfile "demo-service/Dockerfile"
```

> As usual, AWS Copilot generates a `manifest.yml` file that you can use to further customise the resources before you proceed to the actual deployment. To enable VPC access by your application all you need is to uncomment is the following section from the manifest file:

Reviewer: Should we provide the location to the manifest file?
```suggestion
As usual, AWS Copilot generates a `manifest.yml` file that you can use to further customise the resources before you proceed to the actual deployment. To enable VPC access by your application all you need is to uncomment is the following section from the manifest file `copilot/demo-service/manifest.yml`:
```
